homestar-runtime/tests/utils.rs

```diff
@@ -0,0 +1,126 @@
+use anyhow::{Context, Result};
+use assert_cmd::{crate_name, prelude::*};
+#[cfg(not(windows))]
+use nix::{
+    sys::signal::{self, Signal},
+    unistd::Pid,
+};
+use once_cell::sync::Lazy;
+use retry::{delay::Fixed, retry};
+use std::net::{IpAddr, Ipv4Addr, Shutdown, SocketAddr, TcpStream};
+use std::{
+    path::PathBuf,
+    process::{Command, Stdio},
+};
+use sysinfo::{PidExt, ProcessExt, ProcessRefreshKind, RefreshKind, SystemExt};
+
+static BIN: Lazy<PathBuf> = Lazy::new(|| assert_cmd::cargo::cargo_bin(crate_name!()));
+const IPFS: &str = "ipfs";
+
+pub fn startup_ipfs() -> Result<()> {
+    let path = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join(".ipfs");
+    println!("starting ipfs daemon...{}", path.to_str().unwrap());
+    let mut ipfs_daemon = Command::new(IPFS)
+        .args([
+            "--repo-dir",
+            path.to_str().unwrap(),
+            "--offline",
+            "daemon",
+            "--init",
+        ])
+        .stdout(Stdio::piped())
+        .spawn()?;
+
+    // wait for ipfs daemon to start by testing for a connection
+    let socket = SocketAddr::new(IpAddr::V4(Ipv4Addr::LOCALHOST), 5001);
+    let result = retry(Fixed::from_millis(500), || {
+        TcpStream::connect(socket).map(|stream| stream.shutdown(Shutdown::Both))
+    });
+
+    if let Err(err) = result {
+        ipfs_daemon.kill().unwrap();
+        panic!("`ipfs daemon` failed to start: {:?}", err);
+    } else {
+        Ok(())
+    }
+}
+
+pub fn stop_homestar() -> Result<()> {
+    Command::new(BIN.as_os_str())
+        .arg("stop")
+        .stdout(Stdio::null())
+        .stderr(Stdio::null())
+        .status()
+        .context("failed to stop Homestar server")?;
+
+    Ok(())
+}
+
+pub fn stop_ipfs() -> Result<()> {
+    let path = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join(".ipfs");
+    Command::new(IPFS)
+        .args(["--repo-dir", path.to_str().unwrap(), "shutdown"])
+        .stdout(Stdio::null())
+        .stderr(Stdio::null())
+        .status()
+        .context("failed to stop IPFS daemon")?;
+
+    Ok(())
+}
+
+pub fn stop_all_bins() -> Result<()> {
+    let _ = stop_ipfs();
+    let _ = stop_homestar();
+
+    Ok(())
+}
+
+#[cfg(not(windows))]
+pub fn kill_homestar_process_unix() -> Result<()> {
```

Comment:
can remove _unix or _windows below, as the target will pick one (even w/ the same name).

Reply:
Ah yeah, much cleaner. Updated to that. 👌